src/utils/hierarchical_model.py

```diff
@@ -0,0 +1,213 @@
+        axs[index].set_title(f'Dendrogram ({method} linkage)')
+        axs[index].set_xlabel('Sample Index')
+        axs[index].set_ylabel('Distance')
+    
+    # Show the plot and return the dendrogram object
+    plt.show()
+
+
+def plot_thresholded_dendrogram(linkage_matrix: ndarray, distance_threshold: int) -> None:
+    """Plots the dendrogram with a horizontal line at the given threshold.
+
+    Args:
+        linkage_matrix (numpy.ndarray): The linkage matrix of the dataset.
+        threshold (float): The threshold distance to plot the horizontal line at.
+    """
+    # Plot the dendrogram
+    plt.figure(figsize=(10, 7))
+    dendrogram(linkage_matrix)
+    
+    # Set the plot title and axis labels
+    plt.title("Dendrogram with Threshold")
+    
+    # Draw the horizontal line
+    plt.axhline(y=distance_threshold, color='black', linestyle='--')
+    
+    # Show the plot
+    plt.show()
+
+
+def hierarchical_clustering(data: pd.DataFrame, k=3) -> ndarray:
+    """To cluster one-hot encoded data using fcluster from hierarchical clustering algorithms.
+
+    Args:
+        linkage_matrix (ndarray): The linkage matrix of the dataset.
+        k (int): number of clusters.
+        criterion (str, optional): The criterion to use in forming flat clusters. Defaults to 'maxclust'.
+
+    Returns:
+        ndarray: An array of length n. T[i] is the flat cluster number to which original observation i belongs.
+    """
+    return fclusterdata(data, t=k, criterion='maxclust', method='ward')
+
+
+def _map_color_cluster(labels: ndarray) -> pd.Series(str):
+    """Assigns a color to each cluster in the given DataFrame based on their label.
+    Args:
+        df (pd.DataFrame): The DataFrame containing the cluster labels.
+
+    Returns:
+        pd.Series(str): A list of colors, where the i-th element corresponds to the color of the i-th row in the DataFrame.
+    """
+    # Get the unique cluster labels from the DataFrame
+    cluster_labels = np.unique(labels)
+
+    # Assign a color to each cluster label
+    cluster_colors = {label: colors_code[i] for i, label in enumerate(cluster_labels)}
+    
+    # Create a new column in the DataFrame to store the cluster colors
+    return pd.Series(labels).apply(lambda x: cluster_colors[x])
+
+
+def get_transform_PCA(data: pd.DataFrame, dimensions: int) -> ndarray:
+    """Performs principal component analysis (PCA) on a given pandas DataFrame data
+        and returns the transformed data.
+
+    Args:
+        data (pd.DataFrame): The input data to be transformed using PCA.
+        dimensions (int): The number of dimensions to keep after the transformation.
+    Returns:
+        ndarray: A numpy ndarray of shape (n_samples, dimensions) representing the transformed data.
+    """
+    # Perform PCA on the data
+    pca = PCA(n_components=dimensions)
+    return pca.fit_transform(data)
+
+
+def plot_clustering_before_reduction(data: pd.DataFrame, labels:ndarray, dimensions: int) -> None:
```

Comment:
Maybe you can apply some minor changes to this function to get rid of plot_clustering_after_reduction().
Minor changes such as adding an optional argument to check if the data need a dimensionality reduction before plotting or not.

Reply:
Thank you, I said I'd do it but then forgot... I grouped them under the function name "plot clustering".